### PR TITLE
Sharded queues

### PR DESCRIPTION
**Pull request by ftschirpke**

### CODE REVIEW

**Review by ftschirpke — COMMENTED** [review]
When reading it, everything (unless commented) looks logically correct. I will try to test it later.

File: src/bc-pqp-ebpf-kernel.c [hunk 1]
```diff
@@ -227,32 +430,30 @@ static __u64 try_increment_counter(
...
         log("counter increment: failure");
     }
-    // check lower bound
-    if (occupancy + diff > 0) {
-        __sync_fetch_and_add(&queue->occupancy, diff);
-    } else if (occupancy > 0) {
-        __sync_fetch_and_sub(&queue->occupancy, occupancy);
-    }
-    log("occ: %li, pkt: %lu", occupancy, packet_size);
+        queue->occupancy += diff;
```
[the remark this answers is not in the document]
Comment (corite) [reply src/bc-pqp-ebpf-kernel.c:453]: I think we should stay as close to the implementation that is currently in `main` as possible (i.e. I wouldn't pay too much attention to the implementation here). Since there may still come some changes to `burst_control` in the coming days (haven't looked into it yet) I would postpone merging the `main` changes in until then.

File: src/bc-pqp-ebpf-kernel.c [hunk 2]
```diff
@@ -134,16 +152,191 @@ struct phantom_queue {
...
+    __type(value, struct global_queues);
+    __uint(max_entries, 1);
+    __uint(pinning, LIBBPF_PIN_BY_NAME);
+} global_queues_map SEC(".maps");
+
+
+static __u32 timer_callback(
+    void* map, __u32* key, struct global_queues* globals
+) {
+    __u64 aggregate_throughput[PHANTOM_QUEUES];
```
[the remark this answers is not in the document]
Comment (corite) [reply src/bc-pqp-ebpf-kernel.c:194]: Yeah, the clang flag does that. I know, I added the clang flag mainly for the reason that without it our program will be rejected by the loader if you set `PHANTOM_QUEUES=1` for reasons I don't understand. Adding the flag somehow fixed that, and then we also don't need to explicitly initialize that array. But it's kind of a dirty hack I think so I'll add the explicit initialization back in, although strictly speaking we don't need it in our setup.

File: src/bc-pqp-ebpf-kernel.c [hunk 3]
```diff
@@ -134,16 +152,191 @@ struct phantom_queue {
...
+struct {
+    __uint(type, BPF_MAP_TYPE_ARRAY);
+    __type(key, __u32);
+    __type(value, struct global_queues);
+    __uint(max_entries, 1);
+    __uint(pinning, LIBBPF_PIN_BY_NAME);
+} global_queues_map SEC(".maps");
+
+
+static __u32 timer_callback(
```
[the remark this answers is not in the document]
Comment (corite) [reply src/bc-pqp-ebpf-kernel.c:191]: I think so, but I couldn't find a source just now. I think it shouldn't matter though. The goto's are really just there to make the loader happy, since we always stay in bounds of our array those calls should never fail. As for synchronization issues: We only ever read/write (from the local queues) `throughput`, `rate` and `capacity`.

- `throughput` is read and reset by the global timer and is incremented atomically by the local queue.
- `rate`/`capacity` are read and written to by the global timer (and the initialization call, but that happens strictly before everything else) and the local queues only read from those fields.

So I think we can't run into synchronization issues there. The worst that could happen is that an update is delayed, but we could even prevent that with the same strategy as in `initialization`.

Side-note: A possible further performance improvement might be to no longer atomically increment `throughput` and make each local queue do their own reset. For this we'd need another local field though, and I don't know if atomics (without contention) are even slower than performing an additional (non-atomic) read.

File: src/bc-pqp-ebpf-kernel.c [hunk 4]
```diff
@@ -152,24 +345,30 @@ struct {
...
+        res = rate * (__u64)timespan;
+    } else {
+        log("overflowing drain detected (timespan: %ld, rate: %lu)", timespan,
+            rate);
+        res = LONG_MAX;
     }
-    // res is now unsigned again since 0 <= res <= MAX_INT
-    return (__u64)res / ONE_SECOND;
+
+    return res / ONE_SECOND;
```
[the remark this answers is not in the document]
Comment (ftschirpke) [reply src/bc-pqp-ebpf-kernel.c]: Great, thank you.
Yeah, now it technically can still overflow, but only if the timespan becomes one year (~ 2^25) or more while you have a rate of one terabyte/s (~2^40) or more.
So, I personally think, this is a edge case we do not need to handle :sweat_smile:

**Comment by ftschirpke** [comment]
@corite I had quite some issues running the tests for this branch, but now I think this may not be a problem with my tests but maybe with the implementation:
The iperf3 test starts but at some time reports `unable to read from stream socket: Resource temporarily unavailable`.
I now think, this may be due to us not letting any packet through anymore. Could something like this happen?
It's the only explanation I have left at this point.

File: src/bc-pqp-ebpf-kernel.c [hunk 5]
```diff
@@ -152,24 +345,28 @@ struct {
...
     __s64 timespan = (__s64)(now - previous);
     if (timespan < (__s64)0) {
         return 0;
     }
-    __s64 res = timespan * (__s64)rate;
-    if (res < (__s64)0) {
-        res = INT_MAX;
+
+    // we try detect overflows (in a performant fashion)
+    __u64 res;
```
[the remark this answers is not in the document]
Comment (corite) [reply src/bc-pqp-ebpf-kernel.c]: No, I think i just must have forgotten to remove it with the last change (and this branch doesn't have CI yet so I didn't notice).

File: src/bc-pqp-ebpf-kernel.c [hunk 6]
```diff
@@ -156,28 +174,137 @@ struct phantom_queue {
...
+        }
+        bpf_for(ph_queue, 0, PHANTOM_QUEUES) {
+            __u64 curr_tp = lq->queues[ph_queue].throughput;
+            __u64 prev_tp = globals->queues[ph_queue]
+                                .rx_queue_throughput[rx_queue];
+            // reset local throughput counter
+            // (1 to avoid division by 0 problems)
+            __sync_lock_test_and_set(&lq->queues[ph_queue].throughput, 1);
+            // update global moving average
+            __u64 avg = (prev_tp + curr_tp) / 2;
```
Comment (corite) [inline src/bc-pqp-ebpf-kernel.c:241]: Because of the averaging the sum of the local rates might be bigger than the configured global rate. The error should go down over time if the traffic distribution is stable, but still. We should think about whether & how we can avoid this, or maybe also try to disable the averaging altogether (i.e. always only use the last throughput measurement) and see what that does.

Comment (corite) [reply src/bc-pqp-ebpf-kernel.c:241]: I tried disabling the averaging and it doesn't change anything really, so this doesn't seem to be the culprit. When looking at the measurements more closely it seems that the only thing that is really wrong is burst control (or at least _something_ at the start of the flow, it might also be related to the changing configuration of the local queues).

Here is the start of a measurement from the simple test executed on the current state of the branch:
```
-----------------------------------------------------------
Server listening on 5201 (test #5)
-----------------------------------------------------------
Accepted connection from 192.168.101.10, port 41584
[  5] local 192.168.102.10 port 5201 connected to 192.168.101.10 port 12345
[ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
[  5]   0.00-0.10   sec  3.44 MBytes   288 Mbits/sec  0.003 ms  38487/40976 (94%)
[  5]   0.10-0.20   sec  44.8 MBytes  3.76 Gbits/sec  0.003 ms  4182/36645 (11%)
[  5]   0.20-0.30   sec  62.7 MBytes  5.26 Gbits/sec  0.002 ms  842/46234 (1.8%)
[  5]   0.30-0.40   sec  21.8 MBytes  1.83 Gbits/sec  0.010 ms  21584/37388 (58%)
[  5]   0.40-0.50   sec  12.4 MBytes  1.04 Gbits/sec  0.003 ms  27906/36901 (76%)
[  5]   0.50-0.60   sec  12.4 MBytes  1.04 Gbits/sec  0.003 ms  32196/41189 (78%)
```
Interestingly, here the burst isn't quite in the beginning (this was the second execution of the test, in the first execution it was at the beginning) but just after it. Also the burst is so intense that it seems like burst control isn't working at all here.

The good news is that after this initial burst the rate stays very stable at exactly the same rate as our current non-sharded implementation.

Comment (corite) [reply src/bc-pqp-ebpf-kernel.c:240]: The reason for this bug occurred to me just after I wrote this message, and I managed to implement a fix just now.

Basically, when we are coming from a period of no throughput (or just after startup) the local queues are all sized the same. When the test starts, all traffic tends to hit one rx queue though, so that particular queue is sized too small and the others are too big. 

Before we get to correct the queue sizes in the  the local queue will already apply burst control (because there is indeed too much traffic). Because the local queue is pretty small though (1 / RX_QUEUES of what it should be) the amount of magic that is added will also be relatively small since we only fill up this queue.

At some point the timer callback will resize the local queues appropriately, i.e. give the local queue that actually handles all of the traffic more capacity. That means this queue is now allowed to process more, but since it still has the magic in it back from when it was smaller it will never apply burst control again (until the queue is drained once).
